libnetwork/internal/kvstore/boltdb/boltdb.go

```diff
@@ -24,6 +24,7 @@ var (
 
 // BoltDB type implements the Store interface
 type BoltDB struct {
+	mu         sync.Mutex
```

Comment:
Ha! I think I ended up on those changes as well.

I have some changes stashed locally; one of those was that I started to remove the `PersistConnection` bool, because it was always `false`;

> libnetwork/kvstore: remove PersistConnection option
>
> The `PersistConnection` option was added to fix issues arround concurrency.
> By default, the boltdb store would use a persistent connection, and keep the
> database opened, which could cause issues if multiple clients tried to re-use
> the connection, and timed out.
> 
> Libkv commit https://github.com/docker/libkv/commit/b291677af7b0b271bd5ab3c876323f0a7028c226
> switched the default to open and close boltDB connnections for every get/pull
> operation, and added the `PersistConnection` option to allow using the prior
> default (keep a persistent connection).
> 
> The change above was pulled into libnetwork in commit:
> https://github.com/moby/libnetwork/commit/3c623fab9fe6cc737f8d1070bb623b08c5277a73
> 
> And the same PR also removed the timeout through commit:
> https://github.com/moby/libnetwork/commits/bc4190db8793f5a4662bb239b77777b78d60772b
> 
> Removing the timeout resulted in performance issues, and the daemon to hang
> (see libnetwork issues 1374, and moby issue 26251), so a change was merged
> in libkv to allow setting a custom timeout, also if `PersistConnection` was
> not used:
> https://github.com/docker/libkv/commit/63d2c8ad0dcfab3f7034530fa75cdbe6270960df
> 
> The current code in this repository never sets the `PersistConnection`, so
> we can remove support for this option.


_HOWEVER_ after I made that change, and wrote the commit message, I started to scratch my head, and wonder ... _WHY_? Why are we not using a persistent connection? Because a quick look at other consumers of boltdb, seem to be using one connection? So yes, I stashed away that commit as I needed to have a closer look / understanding of the situation 😅